### PR TITLE
Use data size for directory listing cache

### PR DESCRIPTION
**Pull request by NikhilCollooru**
## Description
Use data size for directory listing cache

## Motivation and Context
We currently only have a size control on the number of values but not on the size in bytes. Given this is in memory cache, having a max size in bytes for the cache will be helpful

## Impact
`hive.file-status-cache-size` is deprecated and `hive.file-status-cache.max-retained-size` now represents the max size in bytes for the directory listing cache. 

## Test Plan
existing tests

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Hive Connector Changes
* Add support for setting the max size in bytes for directory listing cache. This can be set via the new `hive.file-status-cache.max-retained-size` configuration property. `hive.file-status-cache-size` is now deprecated :pr:`23176`
```

### CODE REVIEW

File: presto-hdfs-core/src/main/java/com/facebook/presto/hive/BlockLocation.java [hunk 1]
```diff
@@ -83,6 +86,13 @@ public long getLength()
         return length;
     }
 
+    public long getRetainedSizeInBytes()
+    {
+        int hostListSizeInBytes = hosts.stream().map(String::length).reduce(0, Integer::sum);
```
Comment (swapsmagic) [inline presto-hdfs-core/src/main/java/com/facebook/presto/hive/BlockLocation.java:91]: is it possible that this value crosses int limit? If not it's fine else may be convert this into Long as we are doing it for `customSplitInfo` below.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 2]
```diff
@@ -163,7 +164,7 @@ public class HiveClientConfig
     private boolean parquetPushdownFilterEnabled;
     private boolean adaptiveFilterReorderingEnabled = true;
     private Duration fileStatusCacheExpireAfterWrite = new Duration(0, TimeUnit.SECONDS);
-    private long fileStatusCacheMaxSize;
+    private DataSize fileStatusCacheMaxSize = new DataSize(1, GIGABYTE);
```
Comment (swapsmagic) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:167]: Why are we keeping the default value as 1 GB?

File: presto-hdfs-core/src/main/java/com/facebook/presto/hive/HiveFileInfo.java [hunk 3]
```diff
@@ -127,7 +130,15 @@ public Map<String, String> getCustomSplitInfo()
...
-        return path;
+        return new Path(path);
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        long blockLocationsSizeInBytes = blockLocations.stream().map(BlockLocation::getRetainedSizeInBytes).reduce(0L, Long::sum);
+        long extraFileInfoSizeInBytes = extraFileInfo.map(bytes -> bytes.length).orElse(0);
+        long customSplitInfoSizeInBytes = customSplitInfo.entrySet().stream().map(e -> e.getKey().length() + e.getValue().length()).reduce(0, Integer::sum);
+        return INSTANCE_SIZE + path.length() + blockLocationsSizeInBytes + Long.BYTES + Long.BYTES + extraFileInfoSizeInBytes + customSplitInfoSizeInBytes;
```
Comment (vhsu14) [inline presto-hdfs-core/src/main/java/com/facebook/presto/hive/HiveFileInfo.java:141]: Just curious, why do we need 2 Long.BYTES here? I see that there are 2 long fields, are their sizes not included in INSTANCE_SIZE?

File: presto-hdfs-core/src/main/java/com/facebook/presto/hive/BlockLocation.java (hunk 1 above)
Comment (NikhilCollooru) [reply presto-hdfs-core/src/main/java/com/facebook/presto/hive/BlockLocation.java:91]: done. converted to Long.

File: presto-hdfs-core/src/main/java/com/facebook/presto/hive/HiveFileInfo.java (hunk 3 above)
Comment (NikhilCollooru) [reply presto-hdfs-core/src/main/java/com/facebook/presto/hive/HiveFileInfo.java:141]: good catch. INSTANCE_SIZE already includes the size of primitive variables. fixed it.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java (hunk 2 above)
Comment (NikhilCollooru) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: changed it to 0.

**Review by swapsmagic — DISMISSED** [review]

**Review by steveburnett — COMMENTED** [review]
The deprecated configuration property `hive.file-status-cache-size does not appear to be documented, so no doc change is needed for it. 

Suggest add documentation for `hive.file-status-cache.max-retained-size` configuration property - maybe to https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/hive.rst#hive-configuration-properties

**Review by elharo — APPROVED** [review]

**Review by steveburnett — APPROVED** [review]
LGTM! (docs) 

Pull branch, new local doc build, looks good. Thanks!

**Review by elharo — APPROVED** [review]

**Review by feilong-liu — APPROVED** [review]

**Pull request merged**